File: src/Mapping/Mapper.java
package Mapping;

import Utils.Position;
import Utils.ShowMap;

public class Mapper {
    private Map map;
    private ShowMap showmap;

    public Mapper(Map map, ShowMap showmap) {
        this.map = map;
        this.showmap = showmap;
    }

    public void updateMap(Position laserPosition, double[] laserEchoes, double[] laserAngles) {
        Position laserInMap = this.map.toMapPosition(laserPosition);
        for (int i = 0; i < laserEchoes.length; i++) {
            double distance = laserEchoes[i];
            double angle = laserAngles[i];
            Position echoPosition = getEchoPosition(laserPosition, distance, angle);
            Position echoInMap = this.map.toMapPositionNoLimits(echoPosition);
            updateLine(laserInMap, echoInMap, distance);
        }
        this.showmap.updateMap(this.map.grid, laserInMap.getYInt(), laserInMap.getXInt());
    }

    private Position getEchoPosition(Position laserPosition, double distance, double angle) {
        return new Position(laserPosition.getX() + (distance * Math.cos(angle)),
                laserPosition.getY() + (distance * Math.sin(angle)));
    }

    private double bayesianProbability(double occupiedProbablity, double previousProbability) {
        double emptyProbability = 1 - occupiedProbablity;
        double emptyPreviousProbability = 1 - previousProbability;
        return (occupiedProbablity * previousProbability) / (occupiedProbablity * previousProbability + emptyProbability * emptyPreviousProbability);
    }

    /*
     * Updates the values of the grid in a line between the two points, uses Bresenham's line algorithm.
     *
     * @param pos0 First point
     * @param pos1 Second point
     * @param thereIsAnObstacle True if the laser detected an obstacle, false otherwise
     */
    public void updateLine(Position pos0, Position pos1, double distance) {
        boolean thereIsAnObstacle = (distance < Map.MAX_VALUE_LASERS);
        if (Math.abs(pos1.getYInt() - pos0.getYInt()) < Math.abs(pos1.getXInt() - pos0.getXInt())) {
            if (pos0.getXInt() > pos1.getXInt())
                this.updateLineLow(pos1, pos0);
            else
                this.updateLineLow(pos0, pos1);
        } else {
            if (pos0.getYInt() > pos1.getYInt())
                this.updateLineHigh(pos1, pos0);
            else
                this.updateLineHigh(pos0, pos1);
        }
        if (thereIsAnObstacle && pos1.getXInt() >= 0 && pos1.getYInt() >= 0 && pos1.getXInt() < this.map.width && pos1.getYInt() < this.map.height) {
            double occupiedProbability = getOccupiedProbability(distance);
            if (this.map.grid[pos1.getXInt()][pos1.getYInt()] <= 0.7)
                this.map.grid[pos1.getXInt()][pos1.getYInt()] += 0.01;
            this.map.grid[pos1.getXInt()][pos1.getYInt()] = bayesianProbability(occupiedProbability, this.map.grid[pos1.getXInt()][pos1.getYInt()]);
        }
    }

    private void updateLineLow(Position pos0, Position pos1) {
        int dx = pos1.getXInt() - pos0.getXInt();
        int dy = pos1.getYInt() - pos0.getYInt();
        int y_iterator = 1;
        if (dy < 0) {
            y_iterator = -1;
            dy = -dy;
        }
        int D = 2 * dy - dx;
        int y = pos0.getYInt();

        for (int x = pos0.getXInt(); x <= pos1.getXInt(); x++) {
            if (x >= 0 && y >= 0 && x < this.map.width && y < this.map.height) {
                Position pos0InRealWorld = this.map.toRealWorldPosition(pos0);
                Position pos1InRealWorld = this.map.toRealWorldPosition(pos1);
                double distance = Math.sqrt(Math.pow(pos0InRealWorld.getX() - pos1InRealWorld.getX(), 2) + Math.pow(pos0InRealWorld.getY() - pos1InRealWorld.getY(), 2));
                if (distance < MAX_VALUE_LASERS - 10) {
                    double occupiedProbability = getOccupiedProbability(distance);
                    this.map.grid[x][y] = bayesianProbability(1 - occupiedProbability, this.map.grid[x][y]);
                }
            } else {
                break;
            }
            if (D > 0) {
                y+=y_iterator;
                D -= 2 * dx;
            }
            D += 2 * dy;
        }
    }

    private void updateLineHigh(Position pos0, Position pos1) {
        int dx = pos1.getXInt() - pos0.getXInt();
        int dy = pos1.getYInt() - pos0.getYInt();
        int x_iterator = 1;
        if (dx < 0) {
            x_iterator = -1;
            dx = -dx;
        }
        int D = 2 * dx - dy;
        int x = pos0.getXInt();

        for (int y = pos0.getYInt(); y <= pos1.getYInt(); y++) {
            if (x >= 0 && y >= 0 && x < this.map.width && y < this.map.height) {
                Position pos0InRealWorld = this.map.toRealWorldPosition(pos0);
                Position pos1InRealWorld = this.map.toRealWorldPosition(pos1);
                double distance = Math.sqrt(Math.pow(pos0InRealWorld.getX() - pos1InRealWorld.getX(), 2) + Math.pow(pos0InRealWorld.getY() - pos1InRealWorld.getY(), 2));
                if (distance < MAX_VALUE_LASERS - 10) {
                    double occupiedProbability = getOccupiedProbability(distance);
                    this.map.grid[x][y] = bayesianProbability(1 - occupiedProbability, this.map.grid[x][y]);
                }
            } else {
                return;
            }
            if (D > 0) {
                x+=x_iterator;
                D -= 2 * dy;
            }
            D += 2 * dx;
        }
    }

    private double getOccupiedProbability(double distance) {
        return (((Map.MAX_VALUE_LASERS - distance) / Map.MAX_VALUE_LASERS) + 1) / 2 * 0.98;
    }
}

File: python/mapping/mapper.py
from utils.utils import bresenham_line
from math import hypot, cos, sin
from mapping.map import Map
from utils.position import Position

class Mapper:
    def __init__(self, map_to_update, lasers_distance = 0.15, p_max = 0.98, min_increase = 0.01, max_distance = 40):
        self.__map = map_to_update
        self.__lasers_distance = lasers_distance
        self.__max_distance = max_distance
        self.__p_max = p_max
        self.__min_increase = min_increase
        
    def update(self, robot_pos, lasers):
        lasers_pos_x = robot_pos.x + self.__lasers_distance * cos(robot_pos.angle)
        lasers_pos_y = robot_pos.y + self.__lasers_distance * sin(robot_pos.angle)
        lasers_cell = self.__map.to_grid_pos(Position(lasers_pos_x, lasers_pos_y))
        real_lasers_cell = self.__map.to_real_pos(lasers_cell)
        for laser in lasers:
            angle = robot_pos.angle + laser.angle
            laser_hit = Position(lasers_pos_x + laser.echoe * cos(angle), lasers_pos_y + laser.echoe * sin(angle))
            hit_cell = self.__map.to_grid_pos(laser_hit)
            cells = bresenham_line(lasers_cell.x, lasers_cell.y, hit_cell.x, hit_cell.y)
            for cell in cells:
                if self.__map.is_in_bound(cell):
                    if cell.x == hit_cell.x and cell.y == hit_cell.y:
                        if laser.echoe <= self.__max_distance:
                            occupied_probability = self.__occupied_probability(laser.echoe)
                            if not self.__map.grid[hit_cell.x][hit_cell.y] >= 0.7:
                                occupied_probability += self.__min_increase
                            self.__map.grid[hit_cell.x][hit_cell.y] = self.__bayesian_probability(occupied_probability, self.__map.grid[hit_cell.x][hit_cell.y])
                    else:
                        real_cell = self.__map.to_real_pos(cell)
                        distance = hypot(real_cell.x - real_lasers_cell.x, real_cell.y - real_lasers_cell.y)    
                        occupied_probability = self.__occupied_probability(distance)
                        self.__map.grid[hit_cell.x][hit_cell.y] = self.__bayesian_probability(1 - occupied_probability, self.__map.grid[hit_cell.x][hit_cell.y])

    def __bayesian_probability(self, occupied_probability, previous_probabilty):
        empty_probability = 1 - occupied_probability
        empty_previous_probability = 1 - previous_probabilty
        return ((occupied_probability * previous_probabilty) / 
                (occupied_probability * previous_probabilty + empty_probability * empty_previous_probability))
    
    def __occupied_probability(self, distance):
        return ((self.__max_distance - distance) / self.__max_distance) / 2 * self.__p_max
